[FEATURE] Add `module_alias` resolution capabilities

bradennapier opened the pull request:
Hey Y'all! 

So first of all, never used OCAML in my life so excuse any flaws in style.  Mostly just tried to mimic styling I saw done elsewhere as I implemented this feature.  I did add tests and it is definitely fixing the issues I have been having thus far!

So there are a ton of people that use things like `Webpack`, `Rollup`, or something like `babel-plugin-module-resolver` (including Flow team recently I noticed ;-)).  

## The Feature

This adds an option that is very similar to `module.system.node.resolve_dirname` but with slightly different handling.  It is meant to be used in situations where `Webpack`'s `resolve.modules` or `babel-plugin-module-resolver` (as well as multiple other bundlers which offer this type of functionality).  

It basically allows a separated place to define paths that should be aliased but type-checked normally since its actually part of the main codebase in this case. 

> **Note:** aliases are resolved before any `resolve_dirname` configurations.

```
[options]
module.system.node.resolve_alias=shared
module.system.node.resolve_alias=src
```

In the case of something like the image below:

![image](https://user-images.githubusercontent.com/15365418/48621829-a9818900-e959-11e8-86ba-2c4126fd2abd.png)

We can then do the following in `ComponentOne` or `ComponentTwo`:

```
/* @flow */
import A from 'A';
import B from 'B';
import C from 'C';
```

this would then mimic the `Webpack` configuration of:

```javascript
module.exports = {
  //...
  resolve: {
    modules: ['shared', path.resolve(__dirname, 'src'), 'node_modules']
  }
};
```

or the `babel-plugin-module-resolver` configuration of:

```
{
  "plugins": [
    ["module-resolver", {
      "root": ["shared", "./src"]
    }]
  ]
}
```

> Order of precedence would be `C` (shared), `B` (shared), then `A` (./src)

## The Problem This Fixes

The general idea is to remove the need to use relative paths everywhere and anywhere.  At this point I can't even program without it!  So essentially people will do a few possible things:

### Have a Resolution to their Root 

```
module.system.node.resolve_dirname=./src
module.system.node.resolve_dirname=node_modules
```

This way they can import from their root as modules (@see [Webpack Documentation](https://webpack.js.org/configuration/resolve/#resolve-modules)) (@see [babel-plugin-module-resolver Documentation](https://github.com/tleunen/babel-plugin-module-resolver#getting-started))

### Use a customized styling for `node_modules` 

In my case, I generally use a `shared` folder and set it up to resolve the same as `node_modules` would.  This is especially useful in `React` applications as it provides the ability for components to easily share pieces of code at any level.  

While the `resolve_dirname` kind of works better here (with the other example it breaks the code completely), the problem is that when you're building a file it won't work until its been imported from `outside` a shared folder.

https://github.com/bradennapier/flow-modules-bug-repro

This shows the issue when using it this way.

### Why `resolve_dirname` doesn't work for this

The issue with `resolve_dirname` is that doing this will generally break the entire application due to a change in `0.57.0` 

```
New Features:
* Flow will now only check the files in `node_modules/` which are direct or transitive dependencies of the non-`node_modules` code.
```

This is extended to any defined `resolve_dirname` folders even though it's largely used for this purpose.  We have since discovered that using a relative path can fix it but there are still lingering issues in these cases.

### Examples of People having issues that this solves

Just did some quick searches for examples

#5180 
https://github.com/flowtype/flow-for-vscode/issues/289
https://github.com/facebook/flow/issues/4103
https://github.com/facebook/flow/issues/5494
https://github.com/facebook/flow/issues/5819
https://github.com/facebook/flow/issues/5647
https://github.com/facebook/flow/issues/5586
https://github.com/facebook/flow/issues/5385
https://github.com/tleunen/babel-plugin-module-resolver/issues/233
https://github.com/facebook/flow/issues/5778
https://github.com/facebook/flow/issues/5860
https://github.com/facebook/create-react-app/pull/4105
https://github.com/facebook/flow/issues/6966
https://github.com/tleunen/babel-plugin-module-resolver/issues/235


As well as many more if you search through the various flow repos.

## Testing on OSX

If you don't want to build the source and run OSX you can use the binary I built.  There is a compiled binary for OSX located in the repro i linked within a branch: https://github.com/bradennapier/flow-modules-bug-repro/tree/feature/fixed-by-pr-7185/bin

ugeng commented:
Hi @bradennapier! Looks like you've done very useful and highly-demanded task here!
I didn't see your code and have a question regarding of its capabilities in lerna-managed monorepos.
Will it work if I have only one `.flowconfig` in monorepo's root, where I describe all the aliases (say, all my packages have same structure)?
Thanks!

bradennapier commented:
Hmm well theoretically yes it would work fine for a monorepo I do not see why not.  You’d have to use a non relative “src” so it would bubble up in each repo to find the right root path for each but no reason it wouldn’t work great to solve that issue. 

I’d imagine if not then this at least provides the ground work for what would be needed to support it.  I do have a monorepo I’ll test it on at some point when I have time.

ugeng commented:
@bradennapier, sounds wonderful! No doubt in that!
I've mentioned that issue because when I use `name_mapper` like below:
```
module.name_mapper='^containers' ->'src/containers'
```
Flow throws famous 'can not resolve module' error in every package. 

I definitely sure that you predict and avoid such issues in your PR, but it should be noticed though.

Thank you! Can't wait for PR's being approved.

bradennapier commented:
Have you tried the following:

```
module.name_mapper='^containers\/\(.*\)$' -> '<PROJECT_ROOT>/src/containers/\1'
```

Not sure if the substitution is at all necessary but its how all mine are done :-P

ugeng commented:
hi, @bradennapier! I've just tried a new binary in my lerna-managed monorepo with yarn workspaces enabled.
I've tuned my `.flowconfig` like following
![image](https://user-images.githubusercontent.com/3213351/48775558-38dac380-ecef-11e8-9789-d409c549529e.png)
and checked that i'm on new version (currenly i'm on 0.84 because of react-redux typings issue)
![image](https://user-images.githubusercontent.com/3213351/48775656-85be9a00-ecef-11e8-8ca0-0071a4f0fa73.png)

All seems fine.

And it definitely works in my "host" package
![image](https://user-images.githubusercontent.com/3213351/48775957-59574d80-ecf0-11e8-9522-5796b511f9c2.png)

Like you can see, it provides path to module which, as I can assume, tells us that the module resolved.
But in my "library" packages it works a bit differently.
![image](https://user-images.githubusercontent.com/3213351/48776120-c23ec580-ecf0-11e8-97d4-ddeddd3e6281.png)

It does not provides neither module path, nor any module info.. But vscode plugin doesn't complain about it, so it probably works! 😄

PS. In order to simplify webpack config issues I use [`nwb`](https://github.com/insin/nwb) starter tool both in main (hosting) package and in library packages. I don't know how it could impact on situation but probably it could.

PPS. You set version=0.86 but such version already released and have issues with react-redux typings

PPPS. Thank you for the great job you've done. Seems like all almost perfect )

bradennapier commented:
Yeah I leave it out of the List which would define the alias as a “module” to Flow so it probably is the reason it won’t show up there as a module.  

It’s an alias so it’s simply resolving the given alias in a node_module way while preserving standard flow type checking that is removed for modules. 

Thanks for testing!  Glad it works for your needs :)

bradennapier commented:
@jbrown215 any chance i can get a review on this PR?  I think its a really solid feature and i miss it already since i updated to latest and havent yet built my version of it! 

I updated and fixed conflicts with newest version

jbrown215 commented:
Hey @bradennapier:

I really appreciate that you took the time to make this PR. Unfortunately, I'm not a good person to review it. I'll import it and see if anyone else can take a look, but I can't guarantee you someone will get to it.

bradennapier commented:
Thanks @jbrown215 !  Looks like build is failing with my conflict resolutions.  If someone could review the concept and let me know they'd merge it I will absolutely fix it right away.  I probably will anyway but would be nice to have someone with the proper authority or w/e approve the spec.

jbrown215 commented:
I recommend waiting for someone to respond before you put more work into this. I don't want you to keep putting time into something that we may decide is something we don't want or cannot put in the time to support.

bradennapier commented:
Will do, thanks again!

vjpr commented:
> You could also try `module.system.node.resolve_dirname=./src` which should resolve all directories in src as modules if that is a desired effect.

FYI: This means `modules/foo.js` won't work, but `modules/foo/index.js` will, for `import foo from 'modules/foo`.

vjpr commented:
@jbrown215 Can you please review this as discussed [on Twitter](https://twitter.com/flowtype/status/1089056221703950336).

bradennapier commented:
I would note that this part of the codebase has changed a little bit so would need to adjust accordingly, but happy to look into that should the general concept of the feature be approved for further acceptance (which I would personally like to see). 

We do run into it on every project and it's fairly common practice to use `module alias` for webpack setups.

nmote commented on src/services/inference/module/module_js.ml:
```diff
@@ -407,7 +407,14 @@ module Node = struct
+              loc ?resolution_acc dir (spf "%s%s%s" dirname Filename.dir_sep r)
+            )
+          ))
+        else if SSet.mem dir node_modules_containers then
```
This doesn't look right to me. This is checking the same exact thing as the `if` above, so this code path will never be traversed. Am I missing something?

bradennapier replied:
Modified the handling here but absolutely needs a review :-D

nmote commented:
I chatted briefly with @samwgoldman who had some concerns that this might cause problems with rechecks. I'm still looking into this potential issue and I'll have more to say on Monday. I'm optimistic that we can get this merged in some form in the near future.

nmote commented:
I spent some time investigating this today, and I think this should work out. If I'm missing something, hopefully @samwgoldman corrects me.

Basically, the concern was that complicating module resolution in this way could lead us to miss rechecks. Specifically, if `foo.js` depends on `bar.js`, but another `bar.js` elsewhere is added, and due to module resolution rules, `foo.js` now depends on the *new* `bar.js`, we have to make sure to recheck `foo.js` against the newly created `bar.js`.

Fortunately, it looks to me like our logic for tracking this should generalize quite easily. In fact, I don't think this PR will need to do anything special for everything to fall into place.

I could be misunderstanding something (this is my first time looking at this logic), but I'm optimistic that this can be merged shortly. @bradennapier, when you get a chance could you rebase this so that I can play around with it and make sure that everything behaves the way I expect?

bradennapier commented:
Alright so rebased and cleaned up the commits a bit here, confirmed build worked and the feature still operates as expected.

There is one deviation that could come up from the other options.  I largely did not incorporate things this way because it would have made for a much more significant shift in the way things are done (and may not even be possible since our goal is to isolate the `node_modules` resolution from our custom alias resolution).  

In webpack alias you can do something like

```javascript
resolve: [
   'shared',
   'node_modules',
   'backup_modules'
]
```

which would then resolve in the order provided if possible.  in our scenario (which is the only way i have ever even considered using this feature, and the way i believe most people do use it), our custom aliases would also resolve **before** `node_modules` so having a folder locally which shadows a `node_module` would instead resolve our alias.

bradennapier commented:
## Failure Case (Non Critical but Important)

@see [this commit](https://github.com/facebook/flow/pull/7185/commits/961d65533d91766dc299f2fb54cd9df5319ddb69)

So the above example actually led me to finding a failure case in how this should work (and it also fails in the case of using `module_dirname` which was my suspicion all along.  The reason using `./src` in `module_dirname` is a workaround is because of this bug.

Essentially it is not honoring relative paths properly.  I think I found the reason for this when I was putting this PR together but also think it should properly handle `<PROJECT_ROOT>` - a bit beyond me as-to how to do this properly.  I assume its just about calling `expand_project_root_token_to_string` at the right time, just not familiar enough with `OCAML` to fix this case (and the actual relative path case `./`).

nmote commented:
Hey, just to make sure that everyone is on the same page here: I'm planning to review the rest of the PR once @samwgoldman is satisfied with the part that he has started reviewing.

bradennapier commented:
I spoke with Sam and the only thing needed is to do the extra tests which I haven’t had time for due to moving across the country.  Hope to get to it soon.

nmote commented:
Okay, no rush on our end, I just wanted to make sure you weren't waiting for me.

goodmind commented:
@bradennapier are you available for this PR?

davidroeca commented:
Any updates on this PR? The current work-around is to have a messy flow config with a bunch of module name mappers, which becomes cumbersome.

nmote commented:
I haven't heard anything from @bradennapier in some time, so I'm going to close this. Anyone, feel free to ping me on a replacement PR if you are interested in driving this to completion.
